packages/data-broker/src/services/weather/WeatherService.js

```diff
@@ -0,0 +1,236 @@
+  }
+
+  /**
+   * Fetch API data and return in format of events/analytics
+   *
+   * @param EntityType[] entities
+   * @param string type
+   * @param string[] dataElementCodes
+   * @returns {Promise<Object.<results: Object[], metadata: Object>>}
+   * @private
+   */
+  async _getCurrentWeather(entities, apiResultTranslator) {
+    /*
+     * Current weather uses the historic weather API instead of the current weather API
+     * because we want complete data on what happened yesterday (e.g. total rainfall), rather than
+     * live data of what the weather is like at the moment.
+     *
+     * We also need to take into account the Entity's timezone, because Tupaia's server's yesterday
+     * is different from the Entity's yesterday.
+     */
+    const dateRangeByEntityCode = {};
+
+    for (const entity of entities) {
+      const startDate = moment
+        .tz(entity.timezone)
+        .startOf('day')
+        .subtract(1, 'day')
+        .format('YYYY-MM-DD');
+
+      const endDate = moment
+        .tz(entity.timezone)
+        .startOf('day')
+        .format('YYYY-MM-DD');
+
+      dateRangeByEntityCode[entity.code] = {
+        startDate: startDate,
+        endDate: endDate,
+      };
+    }
+
+    const apiResultByEntityCode = await this._getHistoricWeatherDataForEntities(
+      entities,
+      dateRangeByEntityCode,
+    );
+
+    return apiResultTranslator.translate(apiResultByEntityCode);
+  }
+
+  /**
+   * Fetch API data and return in format of events/analytics
+   *
+   * @param EntityType[] entities
+   * @param string type
+   * @param string[] dataElementCodes
+   * @param options
+   * @returns {Promise<Object.<results: Object[], metadata: Object>>}
+   * @private
+   */
+  async _getHistoricWeather(entities, options, apiResultTranslator) {
+    const { startDate, endDate } = this.dateSanitiser.sanitise(options.startDate, options.endDate);
+
+    const dateRangeByEntityCode = {};
+
+    for (const entity of entities) {
+      dateRangeByEntityCode[entity.code] = {
+        startDate: startDate,
+        endDate: endDate,
+      };
+    }
+
+    const apiResultByEntityCode = await this._getHistoricWeatherDataForEntities(
+      entities,
+      dateRangeByEntityCode,
+    );
+
+    return apiResultTranslator.translate(apiResultByEntityCode);
+  }
+
+  /**
+   * @param EntityType[] entities
+   * @param Object.<entityCode: string: Object.<startDate: string, endDate: string> dateRangeByEntityCode
+   * @returns {Promise<Object.<entityCode: string: apiResult: Object>>}
+   * @private
+   */
+  async _getHistoricWeatherDataForEntities(entities, dateRangeByEntityCode) {
+    // Run requests in parallel for performance
+    const getDataForEntity = async entity => {
+      const apiResult = await this.api.historicDaily(
+        entity.pointLatLon().lat,
+        entity.pointLatLon().lon,
+        dateRangeByEntityCode[entity.code].startDate,
+        dateRangeByEntityCode[entity.code].endDate,
```

Comment:
Ah yeah that's what I had before, but even with the same request dates, each entity may have a different local start/end date because of different timezones.

Reply:
Oh of course! That makes total sense